File: 2868-continuous-subarrays/continuous-subarrays.py
class Solution:
    """
    Algorithm: Use a frequency dictionary to store the frequency of every element in the sliding window.
    Begin the sliding window with left=right=0 and run a loop until right < len(nums).
    Inside the loop, first increment freq[nums[right]] += 1 and then check while max(dict) - min(dict) > 2 increment left and decrement freq
    Time complexity: O(n)
    Space complexity: O(1)
    The sorted map stores elements within the current window. Since the difference between any two elements in a valid 
    window cannot exceed 2, the maximum number of unique elements (k) possible in the map at any time is 3. Therefore, the space complexity is constant, O(1).
    """
    def continuousSubarrays(self, nums: List[int]) -> int:
        freq = defaultdict(int)
        left = right = 0
        count = 0

        while right < len(nums):
            freq[nums[right]] += 1

            while max(freq) - min(freq) > 2:
                freq[nums[left]] -= 1
                if freq[nums[left]] == 0:
                    del freq[nums[left]]
                left +=1
            
            count += right - left + 1
            right += 1
        
        return count

File: 2868-continuous-subarrays/continuous-subarrays.cpp
class Solution {
public:

    long long continuousSubarrays(vector<int>& nums) {
        std::unordered_map<int, int> freq;
        int left = 0, right = 0;
        long long int count = 0;

        while (right < nums.size()) {
            // Increase the frequency of the current number
            freq[nums[right]]++;

            // Check if the difference between max and min values in the map exceeds 2
            while (std::max_element(freq.begin(), freq.end(), 
                    [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
                        return a.first < b.first; // Compare keys for max
                    })->first - std::min_element(freq.begin(), freq.end(), 
                    [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
                        return a.first < b.first; // Compare keys for min
                    })->first > 2) {
                // Decrease the frequency of the element at 'left'
                freq[nums[left]]--;
                if (freq[nums[left]] == 0) {
                    freq.erase(nums[left]); // Remove element if frequency becomes 0
                }
                left++;
            }

            // Increment the count for the valid subarray ranges
            count += right - left + 1;
            right++;
        }

        return count;
    }
     
};